snapfaas/src/syscall_server.rs

```diff
@@ -165,652 +118,659 @@ impl SyscallProcessor {
-                            success: false,
-                            fd: wb.fd,
-                            data: Vec::from("Blob doesn't exist"),
-                        }
-                    };
-                    s.send(result.encode_to_vec())?;
-                }
-                Some(SC::FinalizeBlob(fb)) => {
-                    let result = if let Some(mut newblob) = self.create_blobs.remove(&fb.fd) {
-                        let blob = newblob
-                            .write_all(&fb.data)
-                            .and_then(|_| env.blobstore.save(newblob))
-                            .map_err(|e| SyscallProcessorError::Blob(e))?;
-                        syscalls::BlobResponse {
-                            success: true,
-                            fd: fb.fd,
-                            data: Vec::from(blob.name),
-                        }
-                    } else {
-                        syscalls::BlobResponse {
-                            success: false,
-                            fd: fb.fd,
-                            data: Vec::from("Blob doesn't exist"),
-                        }
-                    };
-                    s.send(result.encode_to_vec())?;
-                }
-                Some(SC::OpenBlob(ob)) => {
-                    let result = if let Ok(file) = env.blobstore.open(ob.name) {
+                        let kernel_fd = {
+                            let blobid = self.max_blob_id;
+                            self.max_blob_id += 1;
+                            let blob = self.env.blobstore.open(dg.function.kernel.clone()).expect("open");
+                            self.blobs.insert(blobid, blob);
+                            blobid
+                        };
+                        let function = syscalls::Function {
+                            memory: dg.function.memory as u64,
+                            app_image: app_image_fd,
+                            runtime: runtime_fd,
+                            kernel: kernel_fd,
+                        };
+                        syscalls::Gate { kind: Some(syscalls::gate::Kind::Direct(syscalls::DirectGate {
+                            privilege: Some(dg.privilege.clone().into()),
+                            invoker_integrity_clearance: Some(dg.invoker_integrity_clearance.clone().into()),
+                            declassify: Some(dg.declassify.clone().into()),
+                            function: Some(function),
+                        }))}
+                    },
+                    fs::Gate::Redirect(rd) => syscalls::Gate { kind: Some(syscalls::gate::Kind::Redirect(syscalls::RedirectGate {
+                        privilege: Some(rd.privilege.clone().into()),
+                        invoker_integrity_clearance: Some(rd.invoker_integrity_clearance.clone().into()),
+                        declassify: Some(rd.declassify.clone().into()),
+                        gate: 0,
```

Comment:
So, the `gate` field is unused here and we always put 0, an illegal fd value?

Reply:
Good catch.

This one is a bit tricky. Basically, the `gate` field in the protocol buffer message is a file descriptor (and entry in `self.dents`), while in the RedirectGate struct, it is an `ObjectRef` (which is basically a wrapper around a database key). But at the moment of listing the redirect gate there is not necessarily a directory entry for the refernced gate, and it seems wasteful to allocate one.

So yes, as it stands, `gate` is basically meaningless in this response... which isn't great.

Reply:
I feel we should at least put a comment here. Or, we can separate out the responses for listing a direct gate and for listing a redirect gate.